File: CodingInterviews/DeleteNode.java
class Solution {
    public ListNode deleteNode(ListNode head, int val) {
        if (head == null)
            return head;

        if (head.val == val)
            head = head.next;

        ListNode pre = head, cur = head.next;

        while (cur != null) {
            if (cur.val == val)
                pre.next = cur.next;
            cur = cur.next;
            pre = pre.next;
        }

        return head;
    }
}

File: LeetCode/CodingInterviews/Go/deleteNode.go
package coding

func deleteNode(head *ListNode, val int) *ListNode {
	if head.Val == val {
		return head.Next
	}
	var pre, cur *ListNode = nil, head
	for cur != nil {
		if cur.Val == val {
			pre.Next = cur.Next
			break
		}
		pre = cur
		cur = cur.Next
	}
	return head
}
